### PR TITLE
[UBER-197] Aggregate components properly

### PR DESCRIPTION
**Pull request by wazsone**
# Contribution checklist

## Related issues
https://front.hc.engineering/workbench/platform/tracker/UBER-197
https://front.hc.engineering/workbench/platform/tracker/UBER-223

## Checklist

* [ ] - UI test added to added/changed functionality?
* [ ] - Screenshot is added to PR if applicable ?
* [x] - Does the code work? Check whether function and logic are correct.
* [ ] - Does Changelog.md is updated with changes?
* [ ] - Does the translations are up to date?
* [x] - Does it well tested?
* [x] - Tested for Chrome.
* [x] - Tested for Safari.
* [x] - Go through the changed code looking for typos, TODOs, commented LOCs, debugging pieces of code, etc.
* [x] - Rebase your branch onto master and upstream branch
* [x] - Is there any redundant or duplicate code?
* [x] - Are required links are linked to PR?
* [ ] - Does new code is well documented ?

### CODE REVIEW

File: packages/presentation/package.json [hunk 1]
```diff
@@ -39,6 +39,7 @@
     "@hcengineering/view": "^0.6.6",
     "svelte": "3.55.1",
     "@hcengineering/client": "^0.6.11",
-    "fast-equals": "^2.0.3"
+    "fast-equals": "^2.0.3",
+    "@hcengineering/tracker": "^0.6.8"
```
Comment (haiodo) [inline packages/presentation/package.json]: Presentation is core package and should not know about tracker

File: dev/prod/src/platform.ts [hunk 2]
```diff
@@ -152,6 +152,7 @@ export async function configurePlatform() {
   addLocation(documentId, () => import(/* webpackChunkName: "document" */ '@hcengineering/document-resources'))
   addLocation(bitrixId, () => import(/* webpackChunkName: "bitrix" */ '@hcengineering/bitrix-resources'))
   addLocation(requestId, () => import(/* webpackChunkName: "request" */ '@hcengineering/request-resources'))
+  addLocation(presentationId, () => import(/* webpackChunkName: "presentation" */ '@hcengineering/presentation/src/resources'))
```
Comment (haiodo) [inline dev/prod/src/platform.ts]: Why we need it?

File: packages/presentation/src/component.ts [hunk 3]
```diff
@@ -0,0 +1,278 @@
+//
```
Comment (haiodo) [inline packages/presentation/src/component.ts]: It should be defined in tracker plugin

Comment (haiodo) [reply packages/presentation/src/component.ts]: Also we need to think in future of how to start use of this pipeline only if tracker plugin is loaded, status support could be move to status plugins as well.

File: packages/presentation/src/component.ts [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+  query: DocumentQuery<T>
+  options?: FindOptions<T>
+
+  refresh: () => void
+}
+
+/**
+ * @public
+ */
+export class ComponentMiddleware extends BasePresentationMiddleware implements PresentationMiddleware {
```
Comment (haiodo) [inline packages/presentation/src/component.ts]: Could we unify this pipeline to support all aggregators via mixin to class, as we have for grouping?

File: packages/presentation/src/resources.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import { Resources } from '@hcengineering/platform'
+import { StatusMiddleware } from './status'
+
+export default async (): Promise<Resources> => ({
```
Comment (haiodo) [inline packages/presentation/src/resources.ts]: Presentation should not be loaded as plugin, I think we need to go try unify aggregator pipeline or move status aggregator to view plugin.

File: plugins/tracker-resources/src/plugin.ts [hunk 6]
```diff
@@ -382,6 +393,10 @@ export default mergeIds(trackerId, tracker, {
     SubIssueQuery: '' as ViewQueryAction,
     GetAllPriority: '' as GetAllValuesFunc,
     GetAllComponents: '' as GetAllValuesFunc,
-    GetAllMilestones: '' as GetAllValuesFunc
+    GetAllMilestones: '' as GetAllValuesFunc,
+    GroupByCategoriesFunc: '' as GroupByCategoriesFunc,
```
Comment (haiodo) [inline plugins/tracker-resources/src/plugin.ts]: Should it contain Component in name?

File: plugins/tracker/src/index.ts [hunk 7]
```diff
@@ -309,6 +311,27 @@ export interface Component extends Doc {
...
+/**
+ * @public
+ *
+ * Allow to query for component keys/values.
+ */
+export class ComponentManager {
+  byId: IdMap<Component>
+
+  constructor (readonly components: Component[]) {
+    this.byId = toIdMap(components)
```
Comment (haiodo) [inline plugins/tracker/src/index.ts]: Should we hold onto only _id, name, space, color?

**Review by haiodo — CHANGES_REQUESTED** [review]

File: dev/prod/src/platform.ts (hunk 2 above)
Comment (wazsone) [reply dev/prod/src/platform.ts]: @haiodo to load ```anticrm/packages/presentation/src/resources.ts```

Comment (haiodo) [reply dev/prod/src/platform.ts]: I understand, but we probable could be without it, if generic approach for aggregation will be selected.

File: models/view/src/index.ts [hunk 8]
```diff
@@ -987,6 +1016,21 @@ export function createModel (builder: Builder): void {
...
   })
+
+  builder.mixin(Core.class.Status, core.class.Class, view.mixin.GroupFuncs, {
+    groupByCategories: view.function.GroupByStatusCategoriesFunc,
+    groupValues: view.function.GroupStatusValuesFunc,
+    hasValue: view.function.HasStatusValueFunc
+  })
+
+  builder.mixin(Core.class.Status, core.class.Class, view.mixin.MiddlewareFuncs, {
+    GetManager: view.function.GetStatusManager,
```
Comment (haiodo) [inline models/view/src/index.ts]: It become too complicated, please just have one resource reference.

```js
interface CategoryAggregationView {
  groupByCategory: ()=> ...
  groupValues: ()=> ...
 ... and combine with MiddlewareFuncs
}
```

File: plugins/view/package.json [hunk 9]
```diff
@@ -26,6 +26,7 @@
     "typescript": "^4.3.5"
   },
   "dependencies": {
+    "svelte": "3.55.1",
```
Comment (haiodo) [inline plugins/view/package.json]: Why we need it here?

File: plugins/view/src/index.ts [hunk 10]
```diff
@@ -307,6 +310,76 @@ export interface AllValuesFunc extends Class<Doc> {
...
+
+/**
+ * @public
+ */
+export type GetManager = Resource<<T extends DocManager>(docs: Doc[]) => T>
+
+/**
+ * @public
+ */
+export type GetStore = Resource<<T extends DocManager>() => Writable<T>>
```
Comment (haiodo) [inline plugins/view/src/index.ts]: I do think we need to make it more simple and have only one Mixin with all functionality we need to perform aggregation grouping.

**Review by haiodo — CHANGES_REQUESTED** [review]
A bit more work is required here.

File: plugins/view/package.json (hunk 9 above)
Comment (wazsone) [reply plugins/view/package.json]: @haiodo To define
```
/**
 * @public
 */
export type GetStore = Resource<<T extends DocManager>() => Writable<T>>
```

File: plugins/view/src/index.ts (hunk 10 above)
Comment (haiodo) [reply plugins/view/src/index.ts:345]: Why you need to return store? Please define function's to get your values.

File: plugins/view/package.json (hunk 9 above)
Comment (haiodo) [reply plugins/view/package.json]: Why you need to return store? Please define function's to get your values.

File: plugins/view/src/index.ts [hunk 11]
```diff
@@ -308,6 +310,86 @@ export interface AllValuesFunc extends Class<Doc> {
...
+export type UpdateCustomSorting = <T extends Doc>(
+  finalOptions: FindOptions<T>,
+  attr: AnyAttribute,
+  mgr: DocManager
+) => void
+
+/**
+ * @public
+ */
+export interface AggregationManager {
```
Comment (haiodo) [inline plugins/view/src/index.ts]: Did you plan to always create interfaces like this one?

**Review by haiodo — DISMISSED** [review]

**Review by BykhovDenis — APPROVED** [review]

**Pull request merged**